### ISSUE TITLE
No early stop with custom early_stopping_metric + questions w.r.t. evaluate and save on best epoch

### ISSUE THREAD

**Issue by GillesJ** [ThilinaRajapakse/simpletransformers#1097]
**Describe the bug**
I was implementing a way to evaluate on the best epoch as by default evaluation is done on last epoch.
This is useful in hyperparameter search with early stopping to save disk usage, I only want to keep the best trained epoch of each hyperparameter setting.
My guess is that there is no built-in setting for this, so I have to reloaded the model from `best_model_dir` for evaluation.

While implementing this, a potential bug in early stopping with custom metrics has popped up together with some additional questions:

1.  As per the epoch eval results below, the early_stopping settings are ignored with a custom metric.

2.  Is there a built-in way to call `eval_model()` on best epoch instead off on last epoch?
If not that, is fine, I implemented this by save-reload from `best_model_dir`

3. Is there a way to print epoch evaluation outputs to show custom metrics, or at least `early_stopping_metric`?
`model_args.evaluate_during_training_verbose` does not print this.

**To Reproduce**
```
from simpletransformers.classification import ClassificationArgs, ClassificationModel
from sklearn.metrics import f1_score, precision_score, recall_score

output_dir = './outputs/these_hyperparams'

model_args = ClassificationArgs()

model_args.num_epochs = 16

model_args.use_early_stopping = True
model_args.early_stopping_consider_epoch = True
model_args.early_stopping_delta = 0
model_args.early_stopping_patience = 3
model_args.early_stopping_metric = 'f1_macro'
model_args.early_stopping_metric_minimize = False

model_args.output_dir = output_dir
model_args.best_model_dir = f'{output_dir}/best_model/'
model_args.save_best_model = True
model_args.save_eval_checkpoints = False
model_args.save_model_every_epoch = False
model_args.save_optimizer_and_scheduler = False
model_args.save_steps = -1

model_args.evaluate_during_training = True
model_args.evaluate_during_training_verbose = False # this doesnt actually print metrics
model_args.evaluate_during_training_steps = -1
model_args.evaluate_each_epoch = True
model_args.eval_batch_size = 32

model = ClassificationModel(
    'roberta',
    'roberta-base',
    num_labels=3,
    args=model_args,
)

# Train the model
eval_metrics = {
    'f1_macro': lambda y_true, y_pred: f1_score(y_true, y_pred, average='macro', zero_division=0),
    'p_macro': lambda y_true, y_pred: precision_score(y_true, y_pred, average='macro', zero_division=0),
    'r_macro': lambda y_true, y_pred: recall_score(y_true, y_pred, average='macro', zero_division=0),
    'f1_micro': lambda y_true, y_pred: f1_score(y_true, y_pred, average='micro', zero_division=0),
    'p_micro': lambda y_true, y_pred: precision_score(y_true, y_pred, average='micro', zero_division=0),
    'r_micro': lambda y_true, y_pred: recall_score(y_true, y_pred, average='micro', zero_division=0),
}

model.train_model(df_train,
                  eval_df=df_dev,
                  **eval_metrics
)

# Evaluate the model on dev and holdout test
result, model_outputs, wrong_predictions = model.eval_model(df_dev, **eval_metrics)
print(f'Dev set: {result}')
result, model_outputs, wrong_predictions = model.eval_model(df_test,  **eval_metrics)
print(f'Test set: {result}')
```

Output in `training_progress_scores.csv` (to 3 decimals for clarity):
global_step | mcc | train_loss | eval_loss | f1_macro | p_macro | r_macro | f1_micro | p_micro | r_micro
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
115 | 0.568 | 0.797 | 0.634 | 0.535 | 0.505 | 0.570 | 0.773 | 0.773 | 0.773
230 | 0.584 | 0.756 | 0.558 | 0.569 | 0.711 | 0.582 | 0.786 | 0.786 | 0.786
345 | 0.544 | 0.539 | 0.699 | 0.557 | 0.620 | 0.551 | 0.768 | 0.768 | 0.768
460 | 0.544 | 0.205 | 0.651 | 0.629 (best_eval custom) | 0.636 | 0.623 | 0.758 | 0.758 | 0.758
575 | 0.548 | 0.263 | 0.782 | 0.607 | 0.620 | 0.600 | 0.763 | 0.763 | 0.763
690 | 0.545 | 0.387 | 0.795 | 0.594 | 0.611 | 0.593 | 0.761 | 0.761 | 0.761
805 | 0.548 | 0.177 | 0.842 | 0.591 | 0.604 | 0.591 | 0.762 | 0.762 | 0.762
920 | 0.523 | 0.103 | 1.003 | 0.603 | 0.627 | 0.590 | 0.753 | 0.753 | 0.753
1035 | 0.527 | 0.160 | 0.961 | 0.606 | 0.622 | 0.596 | 0.753 | 0.753 | 0.753
1150 | 0.517 | 0.092 | 1.081 | 0.602 | 0.600 | 0.603 | 0.737 | 0.737 | 0.737
1265 | 0.530 | 0.104 | 1.189 | 0.606 | 0.617 | 0.600 | 0.752 | 0.752 | 0.752
1380 | 0.535 | 0.067 | 1.295 | 0.613 | 0.627 | 0.604 | 0.756 | 0.756 | 0.756
1495 | 0.533 | 0.088 | 1.307 | 0.621 | 0.621 | 0.621 | 0.747 | 0.747 | 0.747
1610 | 0.551 | 0.065 | 1.385 | 0.600 | 0.615 | 0.596 | 0.765 | 0.765 | 0.765
1725 | 0.557 | 0.003 | 1.418 | 0.627 | 0.638 | 0.620 | 0.766 | 0.766 | 0.766
1840 | 0.564 | 0.001 | 1.418 | 0.628 | 0.639 | 0.622 | 0.769 | 0.769 | 0.769

**Expected behavior**
If early stopping worked on custom metric `f1_macro`, the best_eval score is 0.629 @ epoch 4 (global step 460) and it should have stopped after epoch 8 with a patience of 3. 

The issue is potentially that early stopping does not compare to the custom metric but the default `eval_loss` instead, which won't early stop because the direction of optimization is reversed by `model_args.early_stopping_metric_minimize = False`.

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04 Bionic

**Additional context**
Best epoch saving+reloading+eval is useful for hyperparameter search with `early_stopping` where I don't want to include `num_epochs` as a hyperparameter.

**Comment by GillesJ** [comment]
Never mind the bug: I misspelled `model_args.early_stopping_consider_epoch**s**`  😅.
Early stopping is working as intended.

I would still greatly appreciate your thoughts on questions 2. and 3.

**Comment by ThilinaRajapakse** [comment]
2. There isn't a built-in way to do it.
3. It should print the custom metrics as well. If they aren't being printed, then it's not intentional.
